### PR TITLE
256 fix seed

### PR DESCRIPTION
**Pull request by marianovitasari20**

### CODE REVIEW

**Comment by marianovitasari20** [comment]
I’ve merged the latest changes from main and accepted the main version due to conflicts. Since the function I modified in this branch was changed or deleted in main, I now need to reapply my own changes, so I’m working on it again.

File: icenet_mp/model_service.py [hunk 1]
```diff
@@ -19,12 +21,41 @@
 log = logging.getLogger(__name__)
 
 
+class _DeterministicInterpolate:
+    """Monkey-patch F.interpolate to strip antialias=True for deterministic CUDA backward.
```
Comment (jemrobinson) [inline icenet_mp/model_service.py:25]: I'm not sure this is something we want to do. We previously needed the antialias argument to avoid artifacts when resizing.

File: icenet_mp/model_service.py [hunk 2]
```diff
@@ -19,12 +21,41 @@
...
+
 class ModelService:
     def __init__(self, config: DictConfig) -> None:
         """Initialize the model service."""
         self.config_ = config
         if seed := config.get("seed", None):
-            seed_everything(int(seed), workers=True)
+            seed = int(seed)
+            os.environ["PYTHONHASHSEED"] = str(seed)
+            os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"
```
Comment (jemrobinson) [inline icenet_mp/model_service.py:59]: Reference https://docs.nvidia.com/cuda/cublas/ and explain what this means

**Comment by marianovitasari20** [comment]
It’s becoming nondeterministic again after I merge `main` into this branch. I’ve been investigating it since yesterday. I’ve only run it locally and haven’t tried it on Isambard yet.

**Review by jemrobinson — CHANGES_REQUESTED** [review]
I'm not totally convinced that this is a thing we want to be doing.

If we believe that the real performance of a model training has quite a wide distribution of performance (i.e. a large spread of values across any particular metric) then what do we achieve if we manage to make this deterministic? If we want to compare "model configuration A" to "model configuration B" we would still need to run each of these 10-20 times with different seeds in order to convince ourselves that we understand the model uncertainty.

Would we be better working on running model ensembles instead?

**Comment by IFenton** [comment]
> If we want to compare "model configuration A" to "model configuration B" we would still need to run each of these 10-20 times with different seeds in order to convince ourselves that we understand the model uncertainty.

My understanding was that it would allow us to compare two model configurations better, e.g. if model config A always has a lower error than model config B when they are run with the same seed then we can be more confident that model config A is a sensible thing to do. You'd still need to run each pairing a number of times with different seeds to get more of an idea of spread (/do model ensembles), but it seems like it allows a more like-for-like comparison of two configurations. Is that logic wrong @jemrobinson?

> Would we be better working on running model ensembles instead?
This feels like something we should be doing going forward, but as well rather instead.

**Comment by marianovitasari20** [comment]
> I'm not totally convinced that this is a thing we want to be doing.
> 
> If we believe that the real performance of a model training has quite a wide distribution of performance (i.e. a large spread of values across any particular metric) then what do we achieve if we manage to make this deterministic? If we want to compare "model configuration A" to "model configuration B" we would still need to run each of these 10-20 times with different seeds in order to convince ourselves that we understand the model uncertainty.
> 
> Would we be better working on running model ensembles instead?

I think these are two separate things, @jemrobinson. First, we want to fairly investigate why a particular method behaves the way it does, and for that we need a fair, controlled comparison to draw meaningful conclusions. The deterministic mode with a fixed seed is just a tool for that, it doesn't remove uncertainty from the model itself.
Regarding ensembles, we can absolutely still do that later. Adding a seed here doesn't prevent us from running ensemble experiments, the seed is just for controlled comparisons during development. Under normal settings, we wouldn't use a seed at all.

Yes, I’m aware of the effect of removing antialiasing. I thought we could still use this to compare runs in a fully deterministic mode, and we wouldn't need to use a seed under normal settings. I take your point that ensembles could also be used for uncertainty quantification, and that they can be applied under normal settings without seeds later. However, I’m still inclined to keep the deterministic approach, as it makes it easier to compare different configurations, such as methods or loss functions. The purpose is to ensure a fair comparison. We will revert to running without a seed once we have decided which methods (e.g., loss functions or approaches) is better.

**Comment by jemrobinson** [comment]
Hi @marianovitasari20. I guess my underlying point is this: what does comparing two different configurations with the same seed actually tell us? Each one is a single draw from an underlying distribution with large variance, so we can't simply compare e.g. MAE for the two configurations and say that one is performing better than the other.

What's the thing you want to compare that is made easier by this?

**Comment by jemrobinson** [comment]
> My understanding was that it would allow us to compare two model configurations better, e.g. if model config A always has a lower error than model config B when they are run with the same seed then we can be more confident that model config A is a sensible thing to do. You'd still need to run each pairing a number of times with different seeds to get more of an idea of spread (/do model ensembles), but it seems like it allows a more like-for-like comparison of two configurations. Is that logic wrong @jemrobinson?

Yes, if model A always has lower error than model B across N runs, then we can conclude that it's probably performing better. However, I think this applies whether or not you use the same seed.

**Comment by marianovitasari20** [comment]
> > My understanding was that it would allow us to compare two model configurations better, e.g. if model config A always has a lower error than model config B when they are run with the same seed then we can be more confident that model config A is a sensible thing to do. You'd still need to run each pairing a number of times with different seeds to get more of an idea of spread (/do model ensembles), but it seems like it allows a more like-for-like comparison of two configurations. Is that logic wrong @jemrobinson?
> 
> Yes, if model A always has lower error than model B across N runs, then we can conclude that it's probably performing better. However, I think this applies whether or not you use the same seed.

I understand your point, that’s why we want to run A vs. B with the same seed, and then repeat with different seeds. This still gives multiple runs, but in a fairer way.

**Comment by marianovitasari20** [comment]
Just as an example, I ran these experiments using the same seed (while still in fully deterministic mode, before merging main into this branch) to compare the effect of using L1 loss vs. weighted MSE loss. I still need to run multiple experiments with different seeds before drawing conclusions later, but this is a fairer way.
https://github.com/alan-turing-institute/icenet-mp/issues/258

**Comment by jemrobinson** [comment]
I don't quite see why it's fairer to compare different models with the same seed rather than different seeds. They're going to be drawing a different number of random numbers and using them for different purposes, so I think you're still comparing a single sample from a distribution of trained models against a single sample from a different distribution of trained models.

**Comment by marianovitasari20** [comment]
Discussed further during the coworking meeting with @IFenton and @jemrobinson today.

**Comment by marianovitasari20** [comment]
Removing _DeterministicInterpolate, the team agreed not to go with this approach. Keeping a note here for reference: if anyone wants fully deterministic behaviour in future, this function shows one way to do it. It does come with a performance cost, though personally I think it's acceptable for comparison purposes.

**Comment by jemrobinson** [comment]
> Removing _DeterministicInterpolate, the team agreed not to go with this approach. Keeping a note here for reference: if anyone wants fully deterministic behaviour in future, this function shows one way to do it. It does come with a performance cost, though personally I think it's acceptable for comparison purposes.

To be clear - performance cost isn't a problem, but removing the "antialiasing" argument means that the behaviour of the model is changed.

**Comment by marianovitasari20** [comment]
> > Removing _DeterministicInterpolate, the team agreed not to go with this approach. Keeping a note here for reference: if anyone wants fully deterministic behaviour in future, this function shows one way to do it. It does come with a performance cost, though personally I think it's acceptable for comparison purposes.
> 
> To be clear - performance cost isn't a problem, but removing the "antialiasing" argument means that the behaviour of the model is changed.

Sorry for the confusion, by "performance cost" I meant model quality, not compute. Stripping antialias changes the interpolation behaviour which could affect results. So yes, we're agreeing on the same thing.

**Comment by marianovitasari20** [comment]
I've removed it now.

antialias=True → better signal processing, less aliasing, but nondeterministic
antialias=False → slightly worse interpolation quality, but deterministic

In most ML cases, the difference in quality is negligible compared to reproducibility and stable debugging. 

After more testing, I’m actually still not fully convinced we should remove this, but we can merge it for now rather than delay things further due to disagreement.

**Comment by jemrobinson** [comment]
Can you add some plots/metrics here for discussion?

**Comment by marianovitasari20** [comment]
> Can you add some plots/metrics here for discussion?

Sorry, I just realised it can’t write to Isambard due to permission issues, so I can’t access the checkpoints. I’ll come back to this once that’s been sorted out.

**Comment by marianovitasari20** [comment]
https://wandb.ai/turing-seaice/evaluate/runs/wltxre91 (without _DeterministicInterpolate, antialias=True)
https://wandb.ai/turing-seaice/evaluate/runs/70evz1gf (with _DeterministicInterpolate, antialias=False)
comparison: https://wandb.ai/turing-seaice/evaluate/panel/f6pv3rhsg

**Comment by jemrobinson** [comment]
@marianovitasari20 : can you attach a comparison plot to this issue instead of just the W&B links? Also, if the purpose of these plots are to demonstrate the effect of deterministic running, could we see:

- 5 runs using the same seed fully-deterministic
- 5 runs using the same seed but without _DeterministicInterpolate
- 5 runs using different seeds

I think that would help to show that deterministic running is removing or reducing the spread in the models.

**Comment by marianovitasari20** [comment]
**Training plots**
1) 5 runs using the same seed fully-deterministic (seed: 123) --> the results won’t vary, so this is useful for comparison.
image
2) 5 runs using the same seed but without _DeterministicInterpolate (seed: 123)
image
3) 5 runs using different seeds (seeds: 17, 482, 903, 256, 771)
image

**Comment by marianovitasari20** [comment]
**Point 1 (fully-deterministic) vs Point 3 (non-deterministic)**

seed: 17
image

seed: 482
image

seed: 903
image

seed: 256
image

seed: 771
image

**Comment by marianovitasari20** [comment]
additional, 

seed: 771
image

seed: 256
image

**Comment by marianovitasari20** [comment]
Any thoughts on this @jemrobinson @IFenton?

**Comment by jemrobinson** [comment]
Can you show the spread of different seeds in the deterministic case on one graph and the spread in the non-deterministic case on another? No need for additional runs, but just a different way to see the data.

**Comment by marianovitasari20** [comment]
> Can you show the spread of different seeds in the deterministic case on one graph and the spread in the non-deterministic case on another? No need for additional runs, but just a different way to see the data.

thanks @jemrobinson. I reran two of the seeds (without deterministic function) just to show that it keeps changing. Do you mean like this?

with DET:
image

without DET:
image

**Comment by jemrobinson** [comment]
Yes, that's exactly the right thing. I think this confirms that there's quite a lot of spread between different runs that we can only capture by running multiple different seeds.

Given this, is there still something that a fully-deterministic run would be useful for? If there is, perhaps this is something we want to enable/disable in the base config?

File: icenet_mp/model_service.py [hunk 3]
```diff
@@ -184,6 +190,20 @@ def build_trainer(
...
+
+        # Check warn_only survived Lightning's deterministic setup
+        log.debug(
+            "deterministic_algorithms_enabled: %s",
+            torch.are_deterministic_algorithms_enabled(),
+        )
+        log.debug(
+            "warn_only_enabled: %s",
+            torch.is_deterministic_algorithms_warn_only_enabled(),
+        )
```
Comment (jemrobinson) [inline icenet_mp/model_service.py]: I'm a little confused by this. What's the thing that might be overriden if deterministic=True and what are we doing here to reset it?

**Review by jemrobinson — DISMISSED** [review]

File: icenet_mp/model_service.py (hunk 3 above)
Comment (marianovitasari20) [reply icenet_mp/model_service.py]: I wanted both `deterministic=True` and `warn_only=True`, but Lightning's Trainer with `deterministic=True` internally calls `torch.use_deterministic_algorithms(True)` without `warn_only`, overriding our earlier setting. Re-applying it here was my workaround, without it I was getting errors. Not 100% sure this is the cleanest way to do it though.

**Comment by marianovitasari20** [comment]
> Yes, that's exactly the right thing. I think this confirms that there's quite a lot of spread between different runs that we can only capture by running multiple different seeds.
> 
> Given this, is there still something that a fully-deterministic run would be useful for? If there is, perhaps this is something we want to enable/disable in the base config?

Yes, I think deterministic runs are still useful. The key case is when you want to isolate the effect of a specific change (e.g. a new loss function) without seed variance confounding the comparison, a "better" result might just be a lucky seed rather than a genuine improvement. So I'd suggest to keep the deterministic code in place, but making it off by default, only enabled when you want controlled comparisons or need reproducibility for debugging. I can also add a warning in the config noting that enabling deterministic mode will disable antialiasing, so users are aware of the trade-off.

File: icenet_mp/model_service.py (hunk 3 above)
Comment (jemrobinson) [reply icenet_mp/model_service.py]: Can this be replaced by something simpler like:

```
if torch.are_deterministic_algorithms_enabled():
  torch.use_deterministic_algorithms(True, warn_only=True)
```

**Comment by jemrobinson** [comment]
> > Yes, that's exactly the right thing. I think this confirms that there's quite a lot of spread between different runs that we can only capture by running multiple different seeds.
> > Given this, is there still something that a fully-deterministic run would be useful for? If there is, perhaps this is something we want to enable/disable in the base config?
> 
> Yes, I think deterministic runs are still useful. The key case is when you want to isolate the effect of a specific change (e.g. a new loss function) without seed variance confounding the comparison, a "better" result might just be a lucky seed rather than a genuine improvement. So I'd suggest to keep the deterministic code in place, but making it off by default, only enabled when you want controlled comparisons or need reproducibility for debugging. I can also add a warning in the config noting that enabling deterministic mode will disable antialiasing, so users are aware of the trade-off.

Fine with me. Do we want to separate "run with a given seed" from "run deterministically" in the config file?

**Comment by marianovitasari20** [comment]
> > > Yes, that's exactly the right thing. I think this confirms that there's quite a lot of spread between different runs that we can only capture by running multiple different seeds.
> > > Given this, is there still something that a fully-deterministic run would be useful for? If there is, perhaps this is something we want to enable/disable in the base config?
> > 
> > 
> > Yes, I think deterministic runs are still useful. The key case is when you want to isolate the effect of a specific change (e.g. a new loss function) without seed variance confounding the comparison, a "better" result might just be a lucky seed rather than a genuine improvement. So I'd suggest to keep the deterministic code in place, but making it off by default, only enabled when you want controlled comparisons or need reproducibility for debugging. I can also add a warning in the config noting that enabling deterministic mode will disable antialiasing, so users are aware of the trade-off.
> 
> Fine with me. Do we want to separate "run with a given seed" from "run deterministically" in the config file?

Do you mean to put the seed and deterministic settings separately in the config, like in the beginning?
like this:
seed: 123
fully_deterministic: True # this will turn off anti-aliasing

**Comment by IFenton** [comment]
> Yes, I think deterministic runs are still useful. The key case is when you want to isolate the effect of a specific change (e.g. a new loss function) without seed variance confounding the comparison, a "better" result might just be a lucky seed rather than a genuine improvement. So I'd suggest to keep the deterministic code in place, but making it off by default, only enabled when you want controlled comparisons or need reproducibility for debugging. I can also add a warning in the config noting that enabling deterministic mode will disable antialiasing, so users are aware of the trade-off.

I agree this sounds like a good solution

**Comment by jemrobinson** [comment]
> Do you mean to put the seed and deterministic settings separately in the config, like in the beginning? like this: seed: 123 fully_deterministic: True # this will turn off anti-aliasing

Yes, but probably grouped together under a heading, so something like:

```
random:
  fully_deterministic: True # this will turn off anti-aliasing
  seed: 123
```

and, in fact, probably best to put these in a new directory under `config/` with a `default.yaml` that doesn't have a seed or deterministic running, and a `deterministic.yaml` that does (see e.g. how `predict` has multiple different configs nested under it).

**Comment by marianovitasari20** [comment]
Now, `seed` and `deterministic` are `null` and `false` by default. If you want to enable them, add the following to your `local.yaml`:

```yaml
random:
  fully_deterministic: true  
  seed: 123
```

File: icenet_mp/config/base.yaml [hunk 4]
```diff
@@ -6,9 +6,8 @@ defaults:
...
   - model: naive_unet_naive
   - predict: sic-icenet-2d
+  - random: default
   - train: default
   - _self_
 
-base_path: ../base
-
-seed: 555
+base_path: /lus/lfs1aip2/projects/public/u6iz/shared/base
```
Comment (IFenton) [inline icenet_mp/config/base.yaml]: Did you need to change this base path, given you change it for base_isambardai.yaml?

**Review by IFenton — APPROVED** [review]
Generally LGTM (and runs as expected). Just a minor comment about the base_path

File: icenet_mp/config/base.yaml (hunk 4 above)
Comment (marianovitasari20) [reply icenet_mp/config/base.yaml]: Thanks @IFenton. Sorry, yes, you’re right. I’ve changed it back to the original one.

**Comment by marianovitasari20** [comment]
Hi @jemrobinson,
Sorry, I dismissed the "changes requested" one (not sure how that works), because I got blocked from merging this, and I had already made changes based on the reviews. Based on our meeting today, we can merge it if everything is fine, and it has already been approved by @IFenton. Let me know if this is not the case. Thank you both for the reviews! :)

**Comment by jemrobinson** [comment]
@marianovitasari20 one approval is definitely enough! Thanks for doing this :)
